src/OrleansRuntime/Streams/QueueBalancer/DeploymentBasedQueueBalancer.cs

```diff
@@ -98,19 +98,33 @@ public class DeploymentBasedQueueBalancer : QueueBalancerBase, ISiloStatusListen
             await NotifyListeners();
         }
 
+        private async Task HandleSiloStatusChange()
+        {
+            while (!this.siloStatusOracle.CurrentStatus.IsTerminating())
+            {
+                try
+                {
+                    this.siloStatusSharedState = await this.siloStatusSharedState.NextAsync;
+                    HandleSiloStatusChange(this.siloStatusSharedState.State);
+                }
```

Comment:
As long as the silo is not terminating, the thread will wait for silo status changes, process them, then wait for the next change.  It can only access the change in the it's own thread.